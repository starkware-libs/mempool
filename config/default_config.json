{
  "components.gateway_component.execute": {
    "description": "The component execution flag.",
    "privacy": "Public",
    "value": true
  },
  "components.mempool_component.execute": {
    "description": "The component execution flag.",
    "privacy": "Public",
    "value": true
  },
  "gateway_config.network_config.ip": {
    "description": "The gateway server ip.",
    "privacy": "Public",
    "value": "0.0.0.0"
  },
  "gateway_config.network_config.port": {
    "description": "The gateway server port.",
    "privacy": "Public",
    "value": 8080
  },
  "gateway_config.stateful_tx_validator_config.chain_info.chain_id": {
    "description": "The chain ID of the StarkNet chain.",
    "privacy": "Public",
    "value": "0x0"
  },
  "gateway_config.stateful_tx_validator_config.chain_info.eth_fee_token_address": {
    "description": "Address of the ETH fee token.",
    "privacy": "Public",
    "value": "0x0"
  },
  "gateway_config.stateful_tx_validator_config.chain_info.strk_fee_token_address": {
    "description": "Address of the STRK fee token.",
    "privacy": "Public",
    "value": "0x0"
  },
  "gateway_config.stateful_tx_validator_config.max_nonce_for_validation_skip": {
    "description": "The maximum nonce for which the validation is skipped.",
    "privacy": "Public",
    "value": "0x0"
  },
  "gateway_config.stateful_tx_validator_config.max_recursion_depth": {
    "description": "The maximum recursion depth allowed in a transaction.",
    "privacy": "Public",
    "value": 0
  },
  "gateway_config.stateful_tx_validator_config.validate_max_n_steps": {
    "description": "The maximum number of steps the validation function is allowed to take.",
    "privacy": "Public",
    "value": 0
  },
  "gateway_config.stateless_tx_validator_config.max_calldata_length": {
    "description": "Validates that a transaction has calldata length less than or equal to this value.",
    "privacy": "Public",
    "value": 0
  },
  "gateway_config.stateless_tx_validator_config.max_signature_length": {
    "description": "Validates that a transaction has signature length less than or equal to this value.",
    "privacy": "Public",
    "value": 0
  },
  "gateway_config.stateless_tx_validator_config.validate_non_zero_l1_gas_fee": {
    "description": "If true, validates that a transaction has non-zero L1 resource bounds.",
    "privacy": "Public",
    "value": false
  },
  "gateway_config.stateless_tx_validator_config.validate_non_zero_l2_gas_fee": {
    "description": "If true, validates that a transaction has non-zero L2 resource bounds.",
    "privacy": "Public",
    "value": false
  },
  "rpc_state_reader_config.json_rpc_version": {
    "description": "The json rpc version.",
    "privacy": "Public",
    "value": ""
  },
  "rpc_state_reader_config.url": {
    "description": "The url of the rpc server.",
    "privacy": "Public",
    "value": ""
  }
}