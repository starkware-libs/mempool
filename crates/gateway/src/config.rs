use std::collections::BTreeMap;
use std::net::IpAddr;

use papyrus_config::dumping::{append_sub_config_name, ser_param, SerializeConfig};
use papyrus_config::{ParamPath, ParamPrivacyInput, SerializedParam};
use serde::{Deserialize, Serialize};
use validator::Validate;

#[derive(Clone, Debug, Serialize, Deserialize, Validate, PartialEq)]
pub struct GatewayConfig {
    pub network_config: GatewayNetworkConfig,
    pub stateless_transaction_validator_config: StatelessTransactionValidatorConfig,
}

impl GatewayConfig {
    pub fn create_for_testing() -> Self {
        Self {
            network_config: GatewayNetworkConfig::create_for_testing(),
            stateless_transaction_validator_config:
                StatelessTransactionValidatorConfig::create_for_testing(),
        }
    }
}

impl SerializeConfig for GatewayConfig {
    fn dump(&self) -> BTreeMap<ParamPath, SerializedParam> {
        vec![
            append_sub_config_name(self.network_config.dump(), "network_config"),
            append_sub_config_name(
                self.stateless_transaction_validator_config.dump(),
                "stateless_transaction_validator_config",
            ),
        ]
        .into_iter()
        .flatten()
        .collect()
    }
}

/// The gateway network connection related configuration.
#[derive(Clone, Debug, Serialize, Deserialize, Validate, PartialEq)]
pub struct GatewayNetworkConfig {
    pub ip: IpAddr,
    pub port: u16,
}

impl GatewayNetworkConfig {
    pub fn create_for_testing() -> Self {
        Self { ip: "0.0.0.0".parse().unwrap(), port: 8080 }
    }
}

impl SerializeConfig for GatewayNetworkConfig {
    fn dump(&self) -> BTreeMap<ParamPath, SerializedParam> {
        BTreeMap::from_iter([
            ser_param(
                "ip",
                &self.ip.to_string(),
                "The gateway server ip.",
                ParamPrivacyInput::Public,
            ),
            ser_param("port", &self.port, "The gateway server port.", ParamPrivacyInput::Public),
        ])
    }
}

impl Default for GatewayNetworkConfig {
    fn default() -> Self {
        Self { ip: "0.0.0.0".parse().unwrap(), port: 8080 }
    }
}

#[derive(Clone, Debug, Default, Serialize, Deserialize, Validate, PartialEq)]
pub struct StatelessTransactionValidatorConfig {
    // If true, validates that the resource bounds are not zero.
    pub validate_non_zero_l1_gas_fee: bool,
    pub validate_non_zero_l2_gas_fee: bool,

    pub max_calldata_length: usize,
    pub max_signature_length: usize,
}

impl StatelessTransactionValidatorConfig {
    pub fn create_for_testing() -> Self {
        Self {
            validate_non_zero_l1_gas_fee: true,
            validate_non_zero_l2_gas_fee: false,
            max_calldata_length: 10,
            max_signature_length: 0,
        }
    }
}

impl SerializeConfig for StatelessTransactionValidatorConfig {
    fn dump(&self) -> BTreeMap<ParamPath, SerializedParam> {
        BTreeMap::from_iter([
            ser_param(
                "validate_non_zero_l1_gas_fee",
                &self.validate_non_zero_l1_gas_fee,
                "If true, validates that a transaction has non-zero L1 resource bounds.",
                ParamPrivacyInput::Public,
            ),
            ser_param(
                "validate_non_zero_l2_gas_fee",
                &self.validate_non_zero_l2_gas_fee,
                "If true, validates that a transaction has non-zero L2 resource bounds.",
                ParamPrivacyInput::Public,
            ),
            ser_param(
                "max_signature_length",
                &self.max_signature_length,
                "Validates that a transaction has calldata length less than or equal to this \
                 value.",
                ParamPrivacyInput::Public,
            ),
            ser_param(
                "max_calldata_length",
                &self.max_calldata_length,
                "Validates that a transaction has signature length less than or equal to this \
                 value.",
                ParamPrivacyInput::Public,
            ),
        ])
    }
}
