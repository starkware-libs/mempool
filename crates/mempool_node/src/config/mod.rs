#[cfg(test)]
mod config_test;

use std::collections::BTreeMap;
use std::fs::File;
use std::path::Path;

use clap::Command;
use papyrus_config::dumping::{
    append_sub_config_name,
    ser_param,
    // ser_optional_sub_config,
    // ser_pointer_target_param,
    SerializeConfig,
};
use papyrus_config::loading::load_and_process_config;
use papyrus_config::ParamPrivacyInput;
use papyrus_config::{ConfigError, ParamPath, SerializedParam};
use serde::{Deserialize, Serialize};
use starknet_gateway::GatewayConfig;
use validator::{Validate, ValidationError};

use crate::version::VERSION_FULL;

// The path of the default configuration file, provided as part of the crate.
pub const DEFAULT_CONFIG_PATH: &str = "config/default_config.json";

/// The single crate configuration.
#[derive(Clone, Debug, Serialize, Deserialize, Validate, PartialEq)]
pub struct ModuleExecutionConfig {
    pub execute: bool,
}

impl SerializeConfig for ModuleExecutionConfig {
    fn dump(&self) -> BTreeMap<ParamPath, SerializedParam> {
        BTreeMap::from_iter([ser_param(
            "execute",
            &self.execute,
            "The module execution flag.",
            ParamPrivacyInput::Public,
        )])
    }
}

impl Default for ModuleExecutionConfig {
    fn default() -> Self {
        Self { execute: true }
    }
}

/// The crates configuration.
#[derive(Clone, Debug, Serialize, Deserialize, Validate, PartialEq, Default)]
#[validate(schema(function = "validate_modules_config"))]
pub struct ModulesConfig {
    pub gateway_module: ModuleExecutionConfig,
    pub mempool_module: ModuleExecutionConfig,
}

impl SerializeConfig for ModulesConfig {
    fn dump(&self) -> BTreeMap<ParamPath, SerializedParam> {
        #[allow(unused_mut)]
        let mut sub_configs = vec![
            append_sub_config_name(self.gateway_module.dump(), "gateway_module"),
            append_sub_config_name(self.mempool_module.dump(), "mempool_module"),
        ];

        sub_configs.into_iter().flatten().collect()
    }
}

pub fn validate_modules_config(modules: &ModulesConfig) -> Result<(), ValidationError> {
    if modules.gateway_module.execute || modules.mempool_module.execute {
        return Ok(());
    }

    let mut error = ValidationError::new("Invalid modules configuration.");
    error.message = Some("At least one module should be allowed to execute.".into());
    Err(error)
}

/// The configurations of the various components of the node.
#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Validate, Default)]
pub struct MemPoolNodeConfig {
    #[validate]
    pub modules: ModulesConfig,
    #[validate]
    pub gateway_config: GatewayConfig,
}

impl SerializeConfig for MemPoolNodeConfig {
    fn dump(&self) -> BTreeMap<ParamPath, SerializedParam> {
        #[allow(unused_mut)]
        let mut sub_configs = vec![
            append_sub_config_name(self.modules.dump(), "crates"),
            append_sub_config_name(self.gateway_config.dump(), "gateway_config"),
        ];

        sub_configs.into_iter().flatten().collect()
    }
}

impl MemPoolNodeConfig {
    /// Creates a config object. Selects the values from the default file and from resources with
    /// higher priority.
    fn load_and_process_config_file(
        args: Vec<String>,
        config_file_name: Option<&str>,
    ) -> Result<Self, ConfigError> {
        let config_file_name = match config_file_name {
            Some(file_name) => file_name,
            None => DEFAULT_CONFIG_PATH,
        };

        let default_config_file = File::open(Path::new(config_file_name))?;
        load_and_process_config(default_config_file, node_command(), args)
    }

    pub fn load_and_process(args: Vec<String>) -> Result<Self, ConfigError> {
        Self::load_and_process_config_file(args, None)
    }
    pub fn load_and_process_file(
        args: Vec<String>,
        config_file_name: &str,
    ) -> Result<Self, ConfigError> {
        Self::load_and_process_config_file(args, Some(config_file_name))
    }
}

/// The command line interface of this node.
pub fn node_command() -> Command {
    Command::new("Mempool")
        .version(VERSION_FULL)
        .about("Mempool is a StarkNet mempool node written in Rust.")
}
