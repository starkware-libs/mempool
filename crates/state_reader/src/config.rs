use std::collections::BTreeMap;

use papyrus_config::dumping::{ser_param, SerializeConfig};
use papyrus_config::{ParamPath, ParamPrivacyInput, SerializedParam};
use serde::{Deserialize, Serialize};
use validator::Validate;

#[derive(Clone, Debug, Default, Serialize, Deserialize, Validate, PartialEq)]
pub struct RpcStateReaderConfig {
    pub url: String,
    pub json_rpc_version: String,
}

#[cfg(any(feature = "testing", test))]
impl RpcStateReaderConfig {
    pub fn create_for_testing() -> Self {
        Self { url: "http://localhost:8080".to_string(), json_rpc_version: "2.0".to_string() }
    }
}

impl SerializeConfig for RpcStateReaderConfig {
    fn dump(&self) -> BTreeMap<ParamPath, SerializedParam> {
        BTreeMap::from_iter([
            ser_param("url", &self.url, "The url of the rpc server.", ParamPrivacyInput::Public),
            ser_param(
                "json_rpc_version",
                &self.json_rpc_version,
                "The json rpc version.",
                ParamPrivacyInput::Public,
            ),
        ])
    }
}
