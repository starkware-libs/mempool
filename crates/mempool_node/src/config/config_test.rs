#![allow(unused_imports)]
use crate::config::{node_command, GatewayNetworkConnectionConfig};
use crate::config::{ComponentConfig, ComponentExecutionConfig, MempoolNodeConfig};
use assert_matches::assert_matches;
use papyrus_config::dumping::SerializeConfig;
use papyrus_config::loading::load_and_process_config;
use papyrus_config::presentation::get_config_presentation;
use papyrus_config::validators::ParsedValidationErrors;
use papyrus_config::{SerializationType, SerializedContent, SerializedParam};
use std::env::{self, args};
use std::fs::File;
use std::ops::IndexMut;
use std::path::{Path, PathBuf};
use test_utils::config::{get_config_from_file, test_valid_config_body};
use validator::Validate;

const TEST_FILES_FOLDER: &str = "./src/test_files";
const CONFIG_FILE: &str = "mempool_node_config.json";

fn get_config_file_path() -> PathBuf {
    Path::new(TEST_FILES_FOLDER).join(CONFIG_FILE)
}

#[test]
/// Read the valid config file and validate its content.
fn test_valid_config() {
    let expected_config = MempoolNodeConfig {
        components: ComponentConfig {
            gateway_component: ComponentExecutionConfig { execute: true },
            mempool_component: ComponentExecutionConfig { execute: false },
        },
        gateway_config: GatewayNetworkConnectionConfig {
            ip: "0.0.0.0".parse().unwrap(),
            port: 8080,
        },
    };
    let config_file_path = get_config_file_path();
    let fix = false;
    test_valid_config_body(expected_config, config_file_path, fix);
}

#[test]
fn test_components_config() {
    // Read the valid config file and check that the validator finds no errors.
    let config_file_path = get_config_file_path();
    let mut config =
        get_config_from_file::<MempoolNodeConfig>(config_file_path, node_command()).unwrap();
    assert!(config.validate().is_ok());

    // Invalidate the gateway component and check that the validator finds an error.
    config.components.gateway_component.execute = false;

    assert_matches!(config.validate(), Err(e) => {
        let parse_err = ParsedValidationErrors::from(e);
        let mut error_msg = String::new();
        for error in parse_err.0 {
            if error.param_path == "components.__all__" {
                error_msg.push_str(&error.code);
                break;
            }
        }
        assert_eq!(error_msg, "Invalid components configuration.");
    });

    // Validate the mempool component and check that the validator finds no errors.
    config.components.mempool_component.execute = true;
    assert!(config.validate().is_ok());
}
