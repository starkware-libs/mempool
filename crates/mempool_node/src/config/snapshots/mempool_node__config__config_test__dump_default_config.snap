---
source: crates/mempool_node/src/config/config_test.rs
expression: dumped_default_config
---
{
  "components.gateway_component.execute": {
    "description": "The component execution flag.",
    "value": true,
    "privacy": "Public"
  },
  "components.mempool_component.execute": {
    "description": "The component execution flag.",
    "value": true,
    "privacy": "Public"
  },
  "gateway_config.network_config.ip": {
    "description": "The gateway server ip.",
    "value": "0.0.0.0",
    "privacy": "Public"
  },
  "gateway_config.network_config.port": {
    "description": "The gateway server port.",
    "value": {
      "$serde_json::private::Number": "8080"
    },
    "privacy": "Public"
  },
  "gateway_config.stateful_transaction_validator_config.chain_info.chain_id": {
    "description": "The chain ID of the StarkNet chain.",
    "value": "0x0",
    "privacy": "Public"
  },
  "gateway_config.stateful_transaction_validator_config.chain_info.eth_fee_token_address": {
    "description": "Address of the ETH fee token.",
    "value": "0x0",
    "privacy": "Public"
  },
  "gateway_config.stateful_transaction_validator_config.chain_info.strk_fee_token_address": {
    "description": "Address of the STRK fee token.",
    "value": "0x0",
    "privacy": "Public"
  },
  "gateway_config.stateful_transaction_validator_config.max_nonce_for_validation_skip": {
    "description": "The maximum nonce for which the validation is skipped.",
    "value": "0x0",
    "privacy": "Public"
  },
  "gateway_config.stateful_transaction_validator_config.max_recursion_depth": {
    "description": "The maximum recursion depth allowed in a transaction.",
    "value": {
      "$serde_json::private::Number": "0"
    },
    "privacy": "Public"
  },
  "gateway_config.stateful_transaction_validator_config.validate_max_n_steps": {
    "description": "The maximum number of steps the validation function is allowed to take.",
    "value": {
      "$serde_json::private::Number": "0"
    },
    "privacy": "Public"
  },
  "gateway_config.stateless_transaction_validator_config.max_calldata_length": {
    "description": "Validates that a transaction has signature length less than or equal to this value.",
    "value": {
      "$serde_json::private::Number": "0"
    },
    "privacy": "Public"
  },
  "gateway_config.stateless_transaction_validator_config.max_signature_length": {
    "description": "Validates that a transaction has calldata length less than or equal to this value.",
    "value": {
      "$serde_json::private::Number": "0"
    },
    "privacy": "Public"
  },
  "gateway_config.stateless_transaction_validator_config.validate_non_zero_l1_gas_fee": {
    "description": "If true, validates that a transaction has non-zero L1 resource bounds.",
    "value": false,
    "privacy": "Public"
  },
  "gateway_config.stateless_transaction_validator_config.validate_non_zero_l2_gas_fee": {
    "description": "If true, validates that a transaction has non-zero L2 resource bounds.",
    "value": false,
    "privacy": "Public"
  },
  "rpc_state_reader_config.json_rpc_version": {
    "description": "The json rpc version.",
    "value": "",
    "privacy": "Public"
  },
  "rpc_state_reader_config.url": {
    "description": "The url of the rpc server.",
    "value": "",
    "privacy": "Public"
  }
}
